flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/endpoint/SqlGatewayEndpointFactoryUtils.java

```diff
@@ -0,0 +1,153 @@
+                            SQL_GATEWAY_ENDPOINT_TYPE.key()));
+        }
+        validateSpecifiedEndpointsAreUnique(identifiers);
+
+        List<SqlGatewayEndpoint> endpoints = new ArrayList<>();
+        for (String identifier : identifiers) {
+            final SqlGatewayEndpointFactory factory =
+                    FactoryUtil.discoverFactory(
+                            Thread.currentThread().getContextClassLoader(),
+                            SqlGatewayEndpointFactory.class,
+                            identifier);
+            Configuration endpointConfig =
+                    new DelegatingConfiguration(
+                            configuration,
+                            String.format("%s.%s.", GATEWAY_ENDPOINT_PREFIX, identifier));
+            endpoints.add(
+                    factory.createSqlGatewayEndpoint(
+                            new DefaultSqlGatewayEndpointFactoryContext(service, endpointConfig)));
+        }
+        return endpoints;
+    }
+
+    /**
+     * Creates a utility that helps to validate options for a {@link SqlGatewayEndpointFactory}.
+     *
+     * <p>Note: This utility checks for left-over options in the final step.
+     */
+    public static SqlGatewayEndpointFactoryHelper createSqlGatewayEndpointFactoryHelper(
+            SqlGatewayEndpointFactory endpointFactory, SqlGatewayEndpointFactory.Context context) {
+        return new SqlGatewayEndpointFactoryHelper(endpointFactory, context.getOptions());
+    }
+
+    // ----------------------------------------------------------------------------------------
+
+    /**
+     * Helper utility for validating all options for a {@link SqlGatewayEndpointFactory}.
+     *
+     * @see #createSqlGatewayEndpointFactoryHelper(SqlGatewayEndpointFactory,
+     *     SqlGatewayEndpointFactory.Context)
+     */
+    public static class SqlGatewayEndpointFactoryHelper
+            extends FactoryHelper<SqlGatewayEndpointFactory> {
+
+        private SqlGatewayEndpointFactoryHelper(
+                SqlGatewayEndpointFactory factory, Map<String, String> configOptions) {
+            super(factory, configOptions);
+        }
+    }
+
+    private static class DefaultSqlGatewayEndpointFactoryContext
+            implements SqlGatewayEndpointFactory.Context {
+
+        private final SqlGatewayService service;
+        private final Configuration configuration;
+
+        public DefaultSqlGatewayEndpointFactoryContext(
+                SqlGatewayService service, Configuration configuration) {
+            this.service = service;
+            this.configuration = configuration;
+        }
+
+        @Override
+        public SqlGatewayService getSqlGatewayService() {
+            return service;
+        }
+
+        @Override
+        public ReadableConfig getConfiguration() {
+            return configuration;
+        }
+
+        @Override
+        public Map<String, String> getOptions() {
+            return configuration.toMap();
+        }
```

Comment:
This looks like a design flaw that if `getConfiguration` and `getOptions` are equivalent, then why do we need to provide both of them? 

In DynamicTableFactory, the configuration represents the configs of the session, and options represent the table WITH options. I think we can change the methods to `getFlinkConfiguration()` and `getEndpointOptions()` to make it clear.